Problem (Fix #805): fixed package is not reviewed

yihuang opened the pull request:
Solution:
- Review and vendor-in related code into chain-core and document them.

yihuang commented on chain-core/src/fixed/div.rs:
```diff
@@ -0,0 +1,114 @@
+    let (dh, dl) = hi_lo(d);
+    let (mut q, rr) = (*num / dh, *num % dh);
+    let m = q * dl;
+    *num = up_lo(rr, next_half);
```
This part of code, I don't understand the math behind it.
I think it an approximation, to compute `num / (dh * 2^64 + dl)` based on `num / dh * 2^64 + ?`, an approximation `X`?

tomtau replied:
@lezzokafka

lezzokafka replied:
@yihuang This actually looks like long division to me, where you have the quotient and the remainder part?

yihuang replied:
The q refers to quotient, which is first computed with hi, then adjusted with lo.

devashishdxt commented on chain-core/src/fixed/div.rs:
```diff
@@ -0,0 +1,114 @@
+}
+
+#[inline]
+fn div_overflow(lhs: i128, rhs: i128, frac_nbits: u32) -> (i128, bool) {
```
Isn't `frac_nbits` a constant in our code?

yihuang replied:
Yes, but I think still good to keep the helper functions generic?

devashishdxt replied:
Then, take `nbits` also as a parameter?

yihuang replied:
> Then, take `nbits` also as a parameter?

These codes are just adapted from [fixed](https://gitlab.com/tspiteri/fixed) with minimal changes, so I kept the overall structures.

devashishdxt replied:
Ok.

yihuang replied:
I just made the changes, since it's only related to this single function, so it's small.

devashishdxt commented on chain-core/src/fixed/div.rs:
```diff
@@ -0,0 +1,114 @@
+}
+
+#[inline]
+fn neg_abs2(num: (i128, u128)) -> (bool, (u128, u128)) {
```
Some documentation on what this function does?

yihuang replied:
Done

yihuang commented on chain-core/src/fixed/continued_fraction.rs:
```diff
@@ -70,27 +72,27 @@ pub fn log2(x: Fixed, y: Fixed) -> Fixed {
+    let n3 = staking * (n2 - Fixed::from(1));
     // drop fraction part, no wrap
-    let n4 = n3.checked_to_num::<u64>().unwrap();
+    let n4: u64 = n3.try_into().unwrap();
```
Done

yihuang commented on chain-core/src/fixed/mod.rs:
```diff
@@ -0,0 +1,242 @@
+        $(
+            impl From<$ty> for Fixed {
+                fn from(n: $ty) -> Fixed {
+                    Fixed((n as i128) << FRAC_NBITS)
```
Removed not used implementations, may also help with unit test coverage.

yihuang commented:
> for the code that comes from fixed, it'll be good to acknowledge that in the respective file headers + in the NOTICE file: https://github.com/crypto-com/chain/blob/master/NOTICE

Added the licence in the respective files, added the item in NOTICE file.

tomtau commented on chain-core/src/fixed/display.rs:
```diff
@@ -0,0 +1,445 @@
+// Copyright © 2018–2020 Trevor Spiteri
+
+// This library is free software: you can redistribute it and/or
+// modify it under the terms of either
+//
+//   * the Apache License, Version 2.0 or
+//   * the MIT License
+//
+// at your option.
+//
+// You should have recieved copies of the Apache License and the MIT
+// License along with the library. If not, see
+// <https://www.apache.org/licenses/LICENSE-2.0> and
```
https://github.com/crypto-com/chain/blob/master/client-core/src/hd_wallet/mod.rs#L3

I guess it the redistribution copyright can be licensed under the Apache;
if there were any modifications to the original, also add our modifications copyright to make the license information clear

yihuang replied:
Done

tomtau commented:
bors r+

yihuang commented:
> There are a few merge conflicts. Can you rebase with latest master?

Done

tomtau commented:
bors r+
